pennylane/gradients/pulse_gradient.py

```diff
@@ -152,23 +165,34 @@ def _parshift_and_integrate(
 
     if use_broadcasting:
 
-        def _diff_and_contract(res_list, cjacs, int_prefactor):
-            # This one will not work yet with tuples.
-            diff = (res := qml.math.stack(res_list))[0, 1:-1] - res[1, 1:-1]
-            return qml.math.tensordot(diff, cjacs, axes=[[0], [0]]) * int_prefactor
+        def _psr_and_contract(res_list, cjacs, int_prefactor):
+            # Stack results and slice away the first and last values, corresponding to the initial
+            # condition and the final value of the time evolution, but not to a splitting time
+            res = qml.math.stack(res_list)[:, 1:-1]
+            # Contract the results with the parameter-shift rule coefficients
+            parshift = qml.math.tensordot(psr_coeffs, res, axes=1)
+            return qml.math.tensordot(parshift, cjacs, axes=[[0], [0]]) * int_prefactor
 
     else:
-
-        def _diff_and_contract(res_list, cjacs, int_prefactor):
-            diff = (res := qml.math.stack(res_list))[::2] - res[1::2]
-            return qml.math.tensordot(diff, cjacs, axes=[[0], [0]]) * int_prefactor
+        num_shifts = len(psr_coeffs)
+
+        def _psr_and_contract(res_list, cjacs, int_prefactor):
+            res = qml.math.stack(res_list)
+            # Reshape the results such that the first axis corresponds to the splitting times
+            # and the second axis corresponds to different shifts. All other axes are untouched
+            shape = qml.math.shape(res)
+            new_shape = (shape[0] // num_shifts, num_shifts) + shape[1:]
+            res = qml.math.reshape(res, new_shape)
+            # Contract the results with the parameter-shift rule coefficients
+            parshift = qml.math.tensordot(psr_coeffs, res, axes=[[0], [1]])
+            return qml.math.tensordot(parshift, cjacs, axes=[[0], [0]]) * int_prefactor
```

Comment:
Interesting that the batching case is simpler? What is the reason we don't need to reshape in the broadcasting case?

Reply:
It's because of the ordering of the tapes: With broadcasting, we get all the results that belong to different splitting times $\tau$ in consecutive array entries. Without broadcasting, we get all results for a specific $\tau$ together, then for the next $\tau$ etc...

Reply:
In particular, with broadcasting we get results in 2d `[batch_first_shift, batch_second_shift,...]` where each `batch.._shift` is a 1d array. Without broadcasting we get a flat list `[first_shift_first_tau, second_shift_first_tau..,first_shift_second_tau...]`

Reply:
Thanks :100: